engine/kv_engine.hpp

```diff
@@ -205,14 +205,134 @@ class KVEngine : public Engine {
+  // It returns whatever it finds as is.
+  // It's up to caller to handle the deleted or expired key.
+  template <bool readonly>
+  LookupResult lookupKeyImpl(StringView key) {
+    LookupResult result;
+    HashTable::KeyHashHint hint = hash_table_->GetHint(key);
+    if (readonly) {
+      result.s =
+          hash_table_->SearchForRead(hint, key, PrimaryRecordType,
+                                     &result.entry_ptr, &result.entry, nullptr);
+    } else {
+      result.s = hash_table_->SearchForWrite(hint, key, PrimaryRecordType,
+                                             &result.entry_ptr, &result.entry,
+                                             nullptr);
+    }
+    return result;
+  }
+
+  // Look up the key, checks for expiration or deletion.
+  // If key is deleted or has expired, handle it to ExpireCleaner
+  // Also checks whether the type matches expected type.
+  template <bool owns_lock>
+  LookupResult lookupKey(StringView key, RecordType expected_type) {
+    LookupResult result = lookupKeyImpl<true>(key);
+    if (result.s != Status::Ok) {
+      kvdk_assert(result.s == Status::NotFound, "");
+      return result;
+    }
+    if (result.entry.GetRecordType() == RecordType::StringDeleteRecord) {
+      /// DeleteRecord is already handled by OldRecordCleaner
+      /// TODO: don't use DeleteRecord.
+      /// Use Record with empty value and a specific expire_time(e.g. 0)
+      /// to indicate deletion. Add member function Deleted().
+      /// Let ExpireCleaner handle all expires and deletions.
+      /// TODO: Don't default an expire_time in constructor.
+      /// Set a constant. (e.g. INT_MAX)
+      result.s = Status::NotFound;
+      return result;
+    }
+
+    auto HasExpired = [](HashEntry entry) {
+      switch (entry.GetIndexType()) {
+        case HashIndexType::StringRecord: {
+          kvdk_assert(entry.GetRecordType() == StringDataRecord, "");
+          return entry.GetIndex().string_record->HasExpired();
+        }
+        case HashIndexType::UnorderedCollection:
+        case HashIndexType::Queue:
+        case HashIndexType::Skiplist: {
+          return static_cast<Collection*>(entry.GetIndex().ptr)->HasExpired();
+        }
+        default: {
+          kvdk_assert(false, "Unreachable branch!");
+          std::abort();
+        }
+      }
+    };
+    if (!HasExpired(result.entry)) {
+      result.s = (expected_type & result.entry.GetRecordType())
+                     ? result.s
+                     : Status::WrongType;
+      return result;
+    }
+    // Key expired, lock the slot to do search
+    // Since SetExpireTime() or Delete() must acquire the lock of the slot,
+    // we can safely read or expire the key here.
+    std::unique_lock<SpinMutex> guard{*hash_table_->GetHint(key).spin,
+                                      std::defer_lock};
+    if (!owns_lock) {
+      guard.lock();
+      result = lookupKeyImpl<true>(key);
```

Comment:
why search twice.
In fact, I think we can ignore lock here, let the outside caller do the lock